pkg/authentication/context.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+)
+
+type principalContextKey int
+
+var PrincipalContextKey principalContextKey
+
+type Principal struct {
+	Subject string `json:"sub"`
+}
+
+func NewPrincipalFromClaims(c ReadableClaims) (*Principal, error) {
+	a := new(Principal)
+	if err := c.Claims(a); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func PrincipalContext(ctx context.Context, principal *Principal) context.Context {
+	parent := ctx
+	if ctx == nil {
+		parent = context.Background()
+	}
+
+	return context.WithValue(parent, PrincipalContextKey, principal)
+}
+
+func PrincipalFromContext(ctx context.Context) *Principal {
+	if ctx == nil {
+		return nil
+	}
+
+	return ctx.Value(PrincipalContextKey).(*Principal)
+}
+
+type OAuth2Context struct {
+	client   *oauth2.Config
+	verifier TokenVerifier
+
+	tracer trace.Tracer
+	logger logging.LoggerInterface
+}
+
+func (o *OAuth2Context) LoginRedirect(w http.ResponseWriter, r *http.Request) {
+	_, span := o.tracer.Start(r.Context(), "authentication.OAuth2Context.LoginRedirect")
+	defer span.End()
+
+	additionalInfo, _ := json.Marshal(map[string]string{"referer": r.Referer(), "remote-address": r.RemoteAddr})
```

Comment:
This is because Hydra rejects empty state parameter and needs a certain amount of characters to have enough entropy. I used this two values because it should be more performant than generating pseudorandomic values and because we could actually use those for auditing/logging

Reply:
Oh this is used for the state, I didn't notice. The state needs to be random, I don't think that this is enough.

I had a look at the library we use and it turns out that it does not validate the value of the state on the callback. We need to add that (and we need to start storing the state in the cookie).

To add to this we should also send a random `nonce` to the provider and the ID Token that we get back must have the same nonce parameter.

Reply:
As much as I agree with the use of the nonce parameter to avoid replay attacks, I consider that a nice improvement, but an improvement nonetheless, so I am glad to work on that, just not in this PR since it's definitely out scope.
About the state, hydra makes it mandatory, it usually isn't. I chose those values to satisfy hydra's requirements.
I would actually use the state parameter to provide a CSRF protection layer, which also would be too much for this first integration.

I am open to change this state value to something pseudo randomic though, and I'm about to do it in next push.
Thanks!

Reply:
I changed my mind and implemented the nonce check in this PR, so I'll push the changes as soon as I'm done addressing all other comments. Thanks!

Reply:
In the PR change coming up today you'll see the nonce implementation. Thing is, nonce should be encrypted, but encryption will come with next story. So for now it will be base64 encoded, and a todo is there to track that (it will be addressed very soon so no need for a new issue).